src/ProtectedRoomsConfig.ts

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2019, 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import AwaitLock from 'await-lock';
+import { extractRequestError, LogService, MatrixClient, Permalinks } from "matrix-bot-sdk";
+import { IConfig } from "./config";
+const PROTECTED_ROOMS_EVENT_TYPE = "org.matrix.mjolnir.protected_rooms";
+
+/**
+ * Manages the set of rooms that the user has explicitly asked to be protected.
+ */
+export default class ProtectedRoomsConfig {
+
+    /**
+     * These are rooms that we explictly asked Mjolnir to protect, usually via the `rooms add` command.
+     * They are not all of the rooms that mjolnir is protecting as with `config.protectAllJoinedRooms`.
+     */
+    private explicitlyProtectedRooms = new Set</*room id*/string>();
+    /** This is to prevent clobbering the account data for the protected rooms if several rooms are explicitly protected concurrently. */
+    private accountDataLock = new AwaitLock();
+
+    constructor(private readonly client: MatrixClient) {
+
+    }
+
+    /**
+     * Load any rooms that have been explicitly protected from a Mjolnir config.
+     * Will also ensure we are able to join all of the rooms.
+     * @param config The config to load the rooms from under `config.protectedRooms`.
+     */
+    public async loadProtectedRoomsFromConfig(config: IConfig): Promise<void> {
```

Comment:
If I understand correctly, calling this code after initialization makes no sense, right?

So perhaps it should be called `start()` (or made private and called from a `start()` method)?

Reply:
Also, muuuuch nicer than the previous version, thanks!

Reply:
Hmm Yes that is true, but I don't know, i don't think it makes sense to hide what it does behind `start`